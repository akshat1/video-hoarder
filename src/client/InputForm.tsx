import { Button, CircularProgress, Container, FormControl, FormHelperText, Grid, IconButton, InputAdornment, InputLabel, OutlinedInput, Theme } from "@material-ui/core";
import { ClearOutlined } from "@material-ui/icons";
import { makeStyles } from "@material-ui/styles";
import React, { ChangeEventHandler, FormEventHandler, FunctionComponent, MouseEventHandler } from "react";

// const parseSearch = (search: string): Record<string, string> => {
//   const queryString = search.substr(1);
//   const dict = {};
//   // The following code doesn't handle query params with array values (foo=bar&foo=baz) but that
//   // doesn't matter to us.
//   queryString
//     .split("&")
//     .map(pair => pair.split("="))
//     .forEach(([key, value]) => dict[key] = value);
//   return dict;
// };

interface ClearButtonProps {
  clearURL: MouseEventHandler<HTMLInputElement>;
  busy: boolean;
}
const ClearButton:FunctionComponent<ClearButtonProps> = ({ clearURL, busy }) => (
  <InputAdornment position="end">
    {busy ? <CircularProgress /> : null}
    <IconButton
      aria-label="clear"
      onClick={clearURL}
    >
      <ClearOutlined />
    </IconButton>
  </InputAdornment>
);

const useStyle = makeStyles((theme: Theme) => ({
  root: {},
  submitContainer: {
    textAlign: "center",
  },
  submitButton: {
    [theme.breakpoints.up("md")]: {
      marginTop: "11px",
    },
    [theme.breakpoints.down("xs")]: {
      width: "100%",
    },
  },
}));

interface InputFormProps {
  busy: boolean;
  clearURL: MouseEventHandler<HTMLInputElement>;
  onChange: ChangeEventHandler<HTMLInputElement>
  onSubmit: FormEventHandler<HTMLFormElement>;
  url: string;
  valid: boolean;
}

export const InputForm:FunctionComponent<InputFormProps> = (props) => {
  const {
    busy,
    clearURL,
    onChange,
    onSubmit,
    url,
    valid,
  } = props;
  const classes = useStyle();
  const validationErrorMessage = url && !valid ? "Invalid URL" : null;
  const isSubmitDisabled = !valid;

  const clearButton = <ClearButton clearURL={clearURL} busy={busy}/>
  return (
    <Container className={classes.root}>
      <form onSubmit={onSubmit}>
        <Grid
          alignItems="stretch"
          container
          spacing={3}
        >
          <Grid
            item
            md={10}
            xs={12}
          >
            <FormControl
              fullWidth
              required
              variant="outlined"
            >
              <InputLabel htmlFor="url-input">Enter URL</InputLabel>
              <OutlinedInput
                onChange={onChange}
                endAdornment={clearButton}
                value={url}
                error={Boolean(url && !valid)}
              />
              <FormHelperText id="url-error-text">{validationErrorMessage}</FormHelperText>
            </FormControl>
          </Grid>
          <Grid
            className={classes.submitContainer}
            item
            md
            xs={12}
          >
            <Button
              className={classes.submitButton}
              color="primary"
              disabled={isSubmitDisabled}
              type="submit"
              variant="contained"
            >
              Download
            </Button>
          </Grid>
        </Grid>
      </form>
    </Container>
  );
};
