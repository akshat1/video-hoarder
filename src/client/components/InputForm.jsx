/**
 * This component renders the URL input form and comprises a text input for the URL, a submit
 * button, a clear button, and appropriate error messages for the URL pattern validation.
 *
 * While this component does accept an `initialValue` prop, I don't really foresee a use for it at
 * the moment. The component is meant to send information only outwards through the `onSubmit`
 * prop.
 *
 * @module client/components/InputForm
 */
import { addJob } from "../redux/actions";
import { Button, Container, FormControl, FormHelperText, Grid, IconButton, InputAdornment, InputLabel, OutlinedInput,  } from "@material-ui/core";
import { ClearOutlined } from "@material-ui/icons";
import { makeStyles } from "@material-ui/styles";
import PropTypes from "prop-types";
import React, { useEffect, useState } from "react";
import { connect } from "react-redux";
//  import * as Style from "./InputForm.less";

const useStyle = makeStyles(theme => ({
  submitContainer: {
    textAlign: "center"
  },
  submitButton: {
    [theme.breakpoints.up("md")]: {
      marginTop: "11px",
    },
    [theme.breakpoints.down("xs")]: {
      width: "100%"
    }
  }
}));

const InputPattern = "^(https?:\\/\\/.+)?$";
export const InputForm = ({ initialValue, onSubmit }) => {
  const classes = useStyle();
  const [url, setURL] = useState(initialValue);
  useEffect(() => setURL(initialValue), [initialValue]);
  const onChange = e => setURL(e.currentTarget.value);
  const clearURL = () => setURL("");
  const isInvalid = !new RegExp(InputPattern).test(url);
  const validationErrorMessage = isInvalid ? "Invalid URL" : null;
  const isSubmitDisabled = url ? isInvalid : true;
  console.log(`!!url? ${Boolean(url)} // url? ${url} // isInvalid? ${isInvalid} // isSubmitDisabled? ${isSubmitDisabled}`);
  const onFormSubmit = (e) => {
    e.preventDefault();
    if (!isInvalid) onSubmit(url);
  }

  const clearButton = (
    <InputAdornment position="end">
      <IconButton
        aria-label="clear"
        onClick={clearURL}
      >
        <ClearOutlined />
      </IconButton>
    </InputAdornment>
  );

  return (
    <Container>
      <form onSubmit={onFormSubmit}>
      <Grid container spacing={3} alignItems="stretch">
        <Grid item xs={12} md={10}>
          <FormControl
            fullWidth
            variant="outlined"
            required
            helperText={validationErrorMessage}
          >
            <InputLabel htmlFor="url-input">Enter URL</InputLabel>
            <OutlinedInput
              id="url-input"
              onChange={onChange}
              value={url}
              endAdornment={clearButton}
              labelWidth={90}
              error={isInvalid}
            />
            <FormHelperText id="url-error-text">{validationErrorMessage}</FormHelperText>
          </FormControl>
        </Grid>
        <Grid item md xs={12} className={classes.submitContainer}>
          <Button
            type="submit"
            variant="contained"
            color="primary"
            disabled={isSubmitDisabled}
            className={classes.submitButton}
          >
            Download
          </Button>
        </Grid>
      </Grid>  
      </form>
    </Container>
  );
}

InputForm.propTypes = {
  /**
   * @function
   * @param {string} url
   */
  onSubmit: PropTypes.func.isRequired,
  /** @type {string} */
  initialValue: PropTypes.string,
};

InputForm.defaultProps = {
  initialValue: ""
};

const stateToProps = () => ({});
const dispatchToProps = { onSubmit: addJob };
export default connect(stateToProps, dispatchToProps)(InputForm);
