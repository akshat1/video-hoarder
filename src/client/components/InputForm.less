@import (reference) "../style/vars.less";

.wrapper {
  padding: @space-small @space-small 0 @space-small; // bottom padding gets added by the grid row gap.
  border: 1px solid gray;
}

.form {
  display: grid;
  grid-template-columns: auto;
  grid-column-gap: @space-small;
  grid-row-gap: @space-small;
  grid-template-areas:
    "urlInput clearButton"
    "validationMessage validationMessage"
    "submitButton submitButton";

  @media @desktop {
    grid-template-columns: auto 2.5rem 10rem;
    grid-template-areas:
    "urlInput clearButton submitButton"
    "validationMessage validationMessage validationMessage";
  }
}

.input {
  display: inline-block;
  .font-size-large;
  padding: @space-small;
  &.invalid {
    // border-color: @color-border;
    box-shadow: inset 0 0 6px @color-error;
  }
}

.clear {
  width: 2rem;
  height: 2rem;
  border-radius: 1rem;
  .border-muted;
  .font-size-large;
  box-shadow: 0 0 4px @color-border;
  cursor: pointer;
  place-self: center;
  align-self: center;
  &:hover {
    box-shadow: 0 0 6px @color-border;
  }

  &:active {
    box-shadow: 0 0 3px 2px @color-border;
  }

  &[disabled] {
    box-shadow: none;
    opacity: 0.1;
  }
}

.submit {
  display: inline-block;
  .font-size-large;
  grid-area: submitButton;
}

.errorMessage {
  grid-area: validationMessage
}
