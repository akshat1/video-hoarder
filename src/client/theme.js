import { createMuiTheme } from '@material-ui/core/styles';

const Theme = createMuiTheme({});

export { Theme };
