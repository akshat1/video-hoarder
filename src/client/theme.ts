import { createTheme } from "@material-ui/core/styles";
import _ from "lodash";

export const getTheme:any = _.memoize(darkMode =>
  createTheme({
    palette: {
      type: darkMode ? "dark" : "light",
    },
  }),
);
