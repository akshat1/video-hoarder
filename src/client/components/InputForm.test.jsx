import React from 'react';
import { shallow } from 'enzyme';
import { ThemeProvider } from '@material-ui/styles';
import { getTheme } from '../theme';
import InputForm from './InputForm.jsx';

describe('components/InputForm', () => {
  test('InputForm matches snapshot', () => {
    const instance = (
      <ThemeProvider theme={getTheme()}>
        <InputForm onSubmit={() => 0}/>
      </ThemeProvider>
    );
    const mounted = shallow(instance);
    expect(mounted).toMatchSnapshot();
  });
});
