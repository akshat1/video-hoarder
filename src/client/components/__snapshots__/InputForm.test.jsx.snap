// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/InputForm InputForm matches snapshot 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      id="urlInput"
      onChange={[Function]}
      pattern="^(https?:\\\\/\\\\/.+)?$"
      placeholder="Enter the URL of the video here"
      required={true}
      spellCheck={false}
      title="Enter the URL of the video here"
      type="url"
      value=""
    />
    <div
      onClick={[Function]}
      role="button"
      tabIndex={0}
      title="Clear"
    >
      <span>
        +
      </span>
    </div>
    <button
      disabled={true}
      role="button"
      title="Submit URL for download"
      type="submit"
    >
      Download
    </button>
  </form>
</div>
`;
